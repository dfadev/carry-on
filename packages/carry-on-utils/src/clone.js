/** @format **/
export default function clone(value) {
  if (Array.isArray(value)) return value.slice();

  if (value instanceof Object) return { ...value };

  return value;
}
