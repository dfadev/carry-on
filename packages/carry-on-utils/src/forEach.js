import keys from "./keys";

export function forOwn(object, iteratee) {
  for (let i = 0, ks = keys(object), len = ks.length; i < len; i += 1) {
    const key = ks[i];
    const val = object[key];
    const result = iteratee(val, key);
    if (result === false) break;
  }

  return object;
}

export default function forEach(collection, iteratee) {
  if (Array.isArray(collection)) {
    for (let i = 0, len = collection.length; i < len; i += 1)
      if (iteratee(collection[i], i) === false) break;
  } else {
    forOwn(collection, iteratee);
  }
  return collection;
}
