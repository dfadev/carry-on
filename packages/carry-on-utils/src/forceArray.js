/** @format **/
export default function forceArray(thing) {
  return Array.isArray(thing) ? thing : [thing];
}
