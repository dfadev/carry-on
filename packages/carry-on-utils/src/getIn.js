import toPath from "./toPath";

export function getInA(state, path, def) {
  let current = state;
  for (let i = 0; i < path.length; i += 1) {
    const key = path[i];
    if (
      current === undefined ||
      current === null ||
      typeof current !== "object" ||
      (Array.isArray(current) && Number.isNaN(key))
    ) {
      return def;
    }
    current = current[key];
  }
  return current === undefined ? def : current;
}

export default function getIn(state, complexKey, def) {
  return getInA(state, toPath(complexKey), def);
}
