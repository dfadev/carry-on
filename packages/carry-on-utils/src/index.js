/** @format **/
export { throttle, debounce } from "throttle-debounce";
export { default as debouncePromise } from "debounce-promise";
export { default as isEqual } from "fast-deep-equal";
export { default as mutateMerge } from "./mutateMerge";

export { default as isFunction } from "./isFunction";
export { default as toPath } from "./toPath";
export { default as getIn, getInA } from "./getIn";
export { default as makeCancelable } from "./makeCancelable";
export { default as setIn } from "./setIn";
export { default as mutateSet, mutateSetA } from "./mutateSet";
export * from "./proxyState";
