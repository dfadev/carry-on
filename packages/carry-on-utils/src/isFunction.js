/** @format **/
// is the thing a function
export default function isFunction(thing) {
  return !!(thing && thing.constructor && thing.call && thing.apply);
}
