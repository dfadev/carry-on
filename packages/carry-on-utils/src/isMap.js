export default function isMap(target) {
	return target instanceof Map
}
