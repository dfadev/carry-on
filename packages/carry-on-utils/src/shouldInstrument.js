/** @format **/
// copy-paste from https://github.com/nx-js/observer-util/blob/master/src/builtIns/index.js

// built-in object can not be wrapped by Proxies, or, to be clear - unfreezed
// their methods expect the object instance as the 'this' instead of the Proxy wrapper
// complex objects are wrapped with a Proxy of instrumented methods
// which switch the proxy to the raw object and to add reactive wiring
const collectionHandlers = {
  get: true,
  has: true,
  forEach: true,
  keys: true,
  values: true,
  entries: true,
  size: true,
  [Symbol.iterator]: true
};

const handlers = new Map([
  [Map, collectionHandlers],
  [Set, collectionHandlers],
  [WeakMap, collectionHandlers],
  [WeakSet, collectionHandlers],
  [Object, false],
  [Array, false],
  [Int8Array, false],
  [Uint8Array, false],
  [Uint8ClampedArray, false],
  [Int16Array, false],
  [Uint16Array, false],
  [Int32Array, false],
  [Uint32Array, false],
  [Float32Array, false],
  [Float64Array, false]
]);

/* eslint-disable no-nested-ternary */
/* istanbul ignore next */ const globalObj =
  typeof global !== "undefined"
    ? global
    : typeof self !== "undefined"
      ? self
      : typeof window !== "undefined"
        ? window
        : {};

export function shouldInstrument({ constructor }) {
  const isBuiltIn =
    typeof constructor === "function" &&
    constructor.name in globalObj &&
    globalObj[constructor.name] === constructor;
  return !isBuiltIn || handlers.has(constructor);
}

export const getCollectionHandlers = ({ constructor }) =>
  handlers.get(constructor);
