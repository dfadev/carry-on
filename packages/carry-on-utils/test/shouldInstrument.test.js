import { getCollectionHandlers, shouldInstrument } from "../src/shouldInstrument";

const tests = [
  {
    should: "return true for Map",
    obj: new Map(),
    result: true
  },
  {
    should: "return true for Set",
    obj: new Set(),
    result: true
  },
  {
    should: "return true for WeakMap",
    obj: new WeakMap(),
    result: true
  },
  {
    should: "return true for WeakSet",
    obj: new WeakSet(),
    result: true
  },
  {
    should: "return true for Object",
    obj: new Object(),
    result: true
  },
  {
    should: "return true for Array",
    obj: new Array(),
    result: true
  },
  {
    should: "return true for Int8Array",
    obj: new Int8Array(),
    result: true
  },
  {
    should: "return true for Uint8Array",
    obj: new Uint8Array(),
    result: true
  },
  {
    should: "return true for Uint8ClampedArray",
    obj: new Uint8ClampedArray(),
    result: true
  },
  {
    should: "return true for Int16Array",
    obj: new Int16Array(),
    result: true
  },
  {
    should: "return true for Uint16Array",
    obj: new Uint16Array(),
    result: true
  },
  {
    should: "return true for Int32Array",
    obj: new Int32Array(),
    result: true
  },
  {
    should: "return true for Uint32Array",
    obj: new Uint32Array(),
    result: true
  },
  {
    should: "return true for Float32Array",
    obj: new Float32Array(),
    result: true
  },
  {
    should: "return true for Float64Array",
    obj: new Float64Array(),
    result: true
  },
  {
    should: "return false for number",
    obj: 1,
    result: false
  },
];

describe("shouldInstrument", function() {
  tests.forEach(function(test) {
    it("should " + test.should, function() {
      expect(shouldInstrument(test.obj)).toBe(test.result);
    });
  });
});

it("should call getCollectionHandlers", () => {
  expect(getCollectionHandlers(new Object())).toMatchSnapshot();
  expect(getCollectionHandlers(new Map())).toMatchSnapshot();
});
