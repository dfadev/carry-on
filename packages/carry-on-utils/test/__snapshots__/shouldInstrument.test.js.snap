// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call getCollectionHandlers 1`] = `false`;

exports[`should call getCollectionHandlers 2`] = `
Object {
  "entries": true,
  "forEach": true,
  "get": true,
  "has": true,
  "keys": true,
  "size": true,
  "values": true,
  Symbol(Symbol.iterator): true,
}
`;
